### PR TITLE
feat: connect share to lesson overview - LESQ-193 and LESQ-469

### PR DESCRIPTION
**Pull request by JBR90**
## Description

Music year: 1995

- Adds share buttons ('all' is at top of lesson overview) and on containers
- Refactors some shared types and improved naming for legacy lessons.
- adds tracking and tests for share all button

figma - https://www.figma.com/file/3i0y2PHT6pECaGgvLhoN8z/Lesson-page-July-2023?type=design&node-id=829-60892&mode=dev
ticket - https://www.notion.so/oaknationalacademy/I-want-a-link-to-share-page-from-lesson-overview-page-966a62a629c846908a0503ec8e89ca19

## How to test

1. Go to https://deploy-preview-2096--oak-web-application.netlify.thenational.academy

Go to legacy lesson page - you should be able to share
Go to lesson page - share should be disabled (all button) and not visible on containers

Check that avo and post hog are receiving correct tracking events 
lessonShareStarted
downloadResourceButtonClicked

## Screenshots

How it used to look (delete if n/a):
{screenshots}

How it should now look:
{screenshots}

## Checklist

- [ ] Added or updated tests where appropriate
- [ ] Manually tested across browsers / devices
- [ ] Considered impact on accessibility
- [ ] Design sign-off
- [ ] Approved by product owner
- [ ] Does this PR update a package with a breaking change

### CODE REVIEW

File: src/__tests__/pages/teachers/programmes/[programmeSlug]/units/[unitSlug]/lessons/[lessonSlug].test.tsx [hunk 1]
```diff
@@ -76,6 +90,24 @@ describe("pages/teachers/lessons", () => {
...
+          hasDownloadableResources: false,
+          expired: true,
+        })}
+      />,
+    );
+
+    const shareButton = screen.queryAllByTestId("share-all-button");
+    const shareLabel = screen.queryAllByText("Share activities with pupils");
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
```
Comment (njblore) [inline src/__tests__/pages/teachers/programmes/[programmeSlug]/units/[unitSlug]/lessons/[lessonSlug].test.tsx:106]: I think I'd prefer doing something like `if (shareButton.length === 0) throw new error('cant find any shre buttons) else ...`, so if there are no buttons and it throws an error it's super clear why

File: src/components/DownloadAndShareComponents/downloadsAndShare.types.ts [hunk 2]
```diff
@@ -61,49 +61,78 @@ export type PreselectedDownloadType = Exclude<
   "video"
 >;
 
+export const isPreselectedDownloadType = (
+  preselected: PreselectedDownloadType | PreselectedShareType | null,
+): preselected is PreselectedDownloadType => {
+  return (
+    preselectedDownloadType.safeParse(preselected).success &&
+    preselected !== "video"
```
Comment (njblore) [inline src/components/DownloadAndShareComponents/downloadsAndShare.types.ts:69]: can preselected equal 'video' if it parses as a download type?

File: src/components/DownloadAndShareComponents/downloadsAndShare.types.ts [hunk 3]
```diff
@@ -61,49 +61,78 @@ export type PreselectedDownloadType = Exclude<
...
-  "exit quiz": ["exit-quiz-questions", "exit-quiz-answers"],
-  worksheet: ["worksheet-pdf", "worksheet-pptx"],
-  "additional material": ["supplementary-pdf", "supplementary-docx"],
-  all: "all",
-};
+type CombinedPreselectedTypeMap = Record<
+  PreselectedDownloadType | PreselectedShareType,
+  {
+    downloadType?: DownloadResourceType[] | "all";
+    shareType?: LessonShareSchema["type"][] | "all";
```
Comment (njblore) [inline src/components/DownloadAndShareComponents/downloadsAndShare.types.ts:90]: we have the `ShareResourceType` above 

side note but this type file is feeling a bit wild, could do with a tidy up? not necessarily to do in this pr, but there seems to be a lot of overlap/duplication making it difficult to find the type you need

File: src/components/HeaderLesson/HeaderLessonDesktop.tsx [hunk 4]
```diff
@@ -64,7 +65,10 @@ export const HeaderLessonDesktop: FC<HeaderLessonProps> = (props) => {
                   <P $font={"body-2"}>{pupilLessonOutcome}</P>
                 </Box>
               )}
-              <HeaderDownloadAllButton {...props} />
+              <Flex>
+                <HeaderDownloadAllButton $mr={24} {...props} />
```
Comment (njblore) [inline src/components/HeaderLesson/HeaderLessonDesktop.tsx:69]: can we use flex gap instead of margin

**Review by njblore — CHANGES_REQUESTED** [review]

**Review by k-huggs — APPROVED** [review]
LGTM ✅

**Review by jack-skerrett-oak — CHANGES_REQUESTED** [review]
Seems to work fine other than the starter quiz missing its share link (I saw Tash had mentioned this in her code review)

File: src/components/Lesson/LessonOverview/LessonOverview.page.tsx [hunk 5]
```diff
@@ -290,6 +312,7 @@ export function LessonOverview({ lesson }: LessonOverviewProps) {
                       title={"Exit quiz"}
                       anchorId="exit-quiz"
                       downloadable={true}
+                      shareable={isLegacyLicense && true}
```
[the remark this answers is not in the document]
Comment (JBR90) [reply src/components/Lesson/LessonOverview/LessonOverview.page.tsx:315]: Ha!  True is doing nothing here, must of been left in when I was testing the container.

File: src/components/DownloadAndShareComponents/downloadsAndShare.types.ts (hunk 3 above)
Comment (JBR90) [reply src/components/DownloadAndShareComponents/downloadsAndShare.types.ts]: I've had a go at reducing the duplication, created a schema file and moved some of the functions out of the types file.

**Review by njblore — APPROVED** [review]
Looks good!

**Review by jack-skerrett-oak — APPROVED** [review]
Download fix looks good as well :) Just waiting on those events!

**Pull request merged**